### PR TITLE
[FIX] mail: several slight messaging menu improvements

### PR DESCRIPTION
**Pull request by alexkuhn**
Before this commit, the Messaging Menu had the following issues:

- Header is too large;
- Header color in dark theme is too white;
- Important notifications (= needactions) are not highlighted enough
  compared to muted notifications;
- Notification body and title looked too alike;
  => title should be more visible than content
- Mouse-hovering on notification items breaks background-inherit style of avatar and IM status;
  => side-effect of `list-group-item-action`
- Desktop: Header items slightly move when changing selection;
  => comes from `fw-bolder`
- Desktop: mark as read button was too small and not visible enough;
- Mobile: too many borders, and they are too strong;
  => undesirable dropdown style
- Mobile: Messaging Menu had poor "New message" button link whereas Discuss app has nice "Start a conversation" button;
- Mobile: "Start a conversation" button is not shown on livechat tab;
- Mobile: when menu is open while Discuss app is open in background, it's unclear whether the messaging menu is open or not;
  => missing background highlight on Messaging Menu toggler
- Mobile: State of Mobile Discuss App was not synced properly when using Messaging Menu at the same time;
  => e.g. unselected tab
- Mobile: Discuss app and Messaging Menu were not properly aligned;
  => Messaging Menu did not offset position based on systray navbar height
- Mobile: active navbar item was not highlighted enough, especially in white theme;
- "OdooBot has a suggestion" is shown on all tabs instead of only in "All" tab like "OdooBot has a request";

This commit fixes all the above issues.
Some styling issues were also affecting the Activity Menu, such as borders and highlighted open state. This commit applies these few improvements on the Activity Menu too.

Before
before

After
after-1
after-2

### CODE REVIEW

File: addons/mail/static/src/core/common/discuss_app_model.js [hunk 1]
```diff
@@ -42,8 +42,8 @@ export class DiscussApp extends Record {
         return super.insert(...arguments);
     }
 
-    /** @type {'mailbox'|'all'|'channel'|'chat'|'livechat'} */
-    activeTab = "all";
+    /** @type {'main'|'channel'|'chat'|'livechat'} */
+    activeTab = "main";
```
Comment (phenix-factory) [inline addons/mail/static/src/core/common/discuss_app_model.js:46]: Should you rename this in stable? I know 17.0 is recent but it's still a stable version

File: addons/mail/static/src/core/common/thread_service.js [hunk 2]
```diff
@@ -615,13 +615,12 @@ export class ThreadService {
...
-            ? "mailbox"
-            : ["chat", "group"].includes(thread.type)
-            ? "chat"
-            : "channel";
+        this.store.discuss.activeTab =
+            !this.ui.isSmall || thread.model === "mail.box"
+                ? "main"
+                : ["chat", "group"].includes(thread.type)
+                ? "chat"
+                : "channel";
```
Comment (phenix-factory) [inline addons/mail/static/src/core/common/thread_service.js:623]: Can we please refactor this with something readable?

File: addons/mail/static/src/utils/common/hooks.js [hunk 3]
```diff
@@ -493,3 +493,22 @@ export function useSequential() {
...
+        },
+        get menuClass() {
+            return `p-0 o-mail-DiscussSystray ${
+                ui.isSmall
+                    ? "o-mail-systrayFullscreenDropdownMenu start-0 w-100 mh-100 d-flex flex-column mt-0 border-0 shadow-lg"
+                    : ""
+            }`;
+        },
+    };
+}
```
Comment (phenix-factory) [inline addons/mail/static/src/utils/common/hooks.js:514]: This is component concern, it should not be inside a hook

File: addons/mail/static/src/core/common/store_service.js [hunk 4]
```diff
@@ -370,13 +370,16 @@ export const storeService = {
...
         services.ui.bus.addEventListener("resize", () => {
             if (!services.ui.isSmall) {
-                store.discuss.activeTab = "all";
+                store.discuss.activeTab = "main";
             } else {
-                store.discuss.activeTab = store.discuss.thread?.type ?? "all";
+                store.discuss.activeTab =
+                    (store.discuss.thread?.type !== "mailbox"
+                        ? store.discuss.thread?.type
+                        : null) ?? "main";
```
Comment (phenix-factory) [inline addons/mail/static/src/core/common/store_service.js:382]: This pattern is not great, look like a soup of `?` it's hard to predict the content of `activeTab`.

File: addons/mail/static/src/core/web/messaging_menu.js [hunk 5]
```diff
@@ -80,8 +81,12 @@ export class MessagingMenu extends Component {
...
         return (
             this.threads.length > 0 ||
-            (this.store.failures.length > 0 && this.store.discuss.activeTab === "all") ||
-            (this.notification.permission === "prompt" && this.store.discuss.activeTab === "all")
+            (this.store.failures.length > 0 &&
+                this.store.discuss.activeTab === "main" &&
+                !this.env.inDiscussApp) ||
+            (this.notification.permission === "prompt" &&
+                this.store.discuss.activeTab === "main" &&
+                !this.env.inDiscussApp)
```
Comment (phenix-factory) [inline addons/mail/static/src/core/web/messaging_menu.js:89]: the condition on `!this.env.inDiscussApp` could just be at the end I think, not repeated on each part of the `||`.

File: addons/mail/static/src/core/common/discuss_app_model.js (hunk 1 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/discuss_app_model.js:46]: 17.0 is not really stable, and also the rename is necessary so that `mailbox` and `all` are the same, to solve the syncing issue I mentioned in the commit message

File: addons/mail/static/src/core/common/thread_service.js (hunk 2 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/thread_service.js:623]: This is a fix PR, not code cleaning. There's a few code cleaning done when I feel it helps the fix, this should be more of a bonus than a requirement for now

File: addons/mail/static/src/utils/common/hooks.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/src/utils/common/hooks.js:514]: This is shared between `MessagingMenu` and `ActivityMenu`.
I had to change the exact same LOCs between both components in the template. This helps ensuring they reuse the same logic in their template

File: addons/mail/static/src/core/common/thread_service.js (hunk 2 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/thread_service.js:623]: But you have diff on it, it's a good opportunity

File: addons/mail/static/src/core/common/store_service.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/store_service.js:382]: I know, I made the simplest fix on this LOC I could, without drastically changing these LOCs more heavily.
Functionally it's correct and fixes the issue, which is a higher priority than "clean" code.
IMO this is more than enough maintainable, so if this takes some effort to "clean" it now then this is wasted time for the time being

File: addons/mail/static/src/core/web/messaging_menu.js (hunk 5 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/web/messaging_menu.js:89]: Code sucks, but it can't be put at end because `this.threads.length > 0` should be `true` regardless of `env.inDiscussApp`.

This `activeTab === "main" && !env.inDiscussApp` matches the old "all" and not match old "mailbox".
This is just a way to not consider "previews" in mailbox, as this screen is quite special

File: addons/mail/static/src/utils/common/hooks.js (hunk 3 above)
Comment (phenix-factory) [reply addons/mail/static/src/utils/common/hooks.js:514]: I still think it should not be a hook, it's not holding any meaning full logic. It's just indirection. It should be done in the template of the component it's simpler.
It also create a dependency between those 2 components making it hard to edit them without a risk to break the other. Component should not be interdependent.

Comment (alexkuhn) [reply addons/mail/static/src/utils/common/hooks.js:514]: I made a hook because it feels natural to reuse logic between 2 components with that, without the identity crisis.
Without using a hook that would mean using a lone function or object, which is unnatural with reusing code between components. Also this add another problem: this logic wants to be UI reactive with service `ui`, so it wants to make use of `useState()`. This makes a lot of sense as a hook, not a lone function.

To sum up, I disagree and it's ok to be a hook

File: addons/mail/static/src/core/common/store_service.js (hunk 4 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/store_service.js:382]: I disagree, cleaning code is never a wast of time as it avoid technical debt. And here I feel like you are creating technical debt.
I think this is better: 
```suggestion
store.discuss.activeTab = "main";
if (store.discuss.thread?.type && store.discuss.thread?.type !== "mailbox") {
    store.discuss.activeTab = store.discuss.thread.type;
}
```

File: addons/mail/static/src/utils/common/hooks.js (hunk 3 above)
Comment (phenix-factory) [reply addons/mail/static/src/utils/common/hooks.js:514]: You miss understand me a bit, I also don't want a lone function or anything. css class logic should be done in the component template (or if complex from a component function). 
The issue is that this hook create dependencies between the components: you can't edit the border of one without editing the border of the other. I don't think it's a good thing.

File: addons/mail/static/src/core/common/thread_service.js (hunk 2 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/thread_service.js:623]: If the motivation is you don't like it (which is not necessarily wrong, I get it too that's it's not ugly and there might be better LOCs to do that) but there's no solution proposal or more genuine other concerns, then this is not worth improving in this PR for now

File: addons/mail/static/src/utils/common/hooks.js (hunk 3 above)
Comment (alexkuhn) [reply addons/mail/static/src/utils/common/hooks.js:514]: If we need separation in the future, then we will adapt the code.
For now, the parts of template of `MessagingMenu` and `ActivityMenu` were exactly the same, and they wanted to solve the exact same issue. This is definitely intentional for them to reuse the same code, so I really don't get why it's a problem for them to reuse the same code when it's exactly the intention of this change in this PR (and the current code of this parts of template confirms it)

File: addons/mail/static/src/core/web/messaging_menu.js (hunk 5 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/web/messaging_menu.js:89]: Ho I understand!
Well I think you still can avoid the duplication like this:
```js
    get hasPreviews() {
        return (
            this.threads.length > 0 || (
                (this.store.failures.length > 0 && this.store.discuss.activeTab === "all") ||
                (this.notification.permission === "prompt" && this.store.discuss.activeTab === "all")
                    (this.store.failures.length > 0 &&
                        this.store.discuss.activeTab === "main" &&
                        !this.env.inDiscussApp) ||
                (this.notification.permission === "prompt" &&
                    this.store.discuss.activeTab === "main"
                )
                && !this.env.inDiscussApp)
        );
    }

```

File: addons/mail/static/src/core/common/store_service.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/store_service.js:382]: Cleaning code can be good time investment, it could be a huge waste of time and can do more harm than good.
The tricky part is that it's named "clean", enforcing a positive interpretation of this adjective, but in practice it's a ideological movement to impose a way to code that can make the code worse and maintainability poor.

With that in mind, "clean" code is not always good.

You proposed LOCs are IMO relatively equivalent. Both are ok, there are tradeoffs with both of them.
On one end, the statements are simpler with your solution. On the other end, it repeats `store.discuss.` quite a lot, and the 1st statement is a lie as it assigns a wrong `activeTab` that is overrides just after, making the code harder to understand by just reading at bits of it.

Comment (phenix-factory) [reply addons/mail/static/src/core/common/store_service.js:382]: I believe I converted an hard to read code pattern to a commonly used one making it "clean" because most dev have seen and use it before and understand it in the blink of an eye. 
We are entering bad faith territory now. If you don't see that my proposition is easier to read and have to question yourself about what clean code is, I don't know what to tell you.

File: addons/mail/static/src/core/web/messaging_menu.dark.scss [hunk 6]
```diff
@@ -0,0 +1,5 @@
+.o-mail-MessagingMenu-navbar button {
+    &.o-active, &:hover {
+        background-color: mix($o-gray-200, $o-gray-300);
+    }
+}
```
Comment (seb-odoo) [inline addons/mail/static/src/core/web/messaging_menu.dark.scss:5]: missing new line

File: addons/mail/static/src/core/web/messaging_menu.scss [hunk 7]
```diff
@@ -8,6 +8,8 @@
...
 }
 
-.o-mail-MessagingMenu-navbar {
-    z-index: 1;
-}
+.o-mail-MessagingMenu-navbar button {
+    &.o-active, &:hover {
+        background-color: mix($o-gray-100, $o-gray-200);
+    }
+}
```
Comment (seb-odoo) [inline addons/mail/static/src/core/web/messaging_menu.scss:15]: missing new line

File: addons/mail/static/src/core/web/notification_item.dark.scss [hunk 8]
```diff
@@ -0,0 +1,9 @@
+
```
Comment (seb-odoo) [inline addons/mail/static/src/core/web/notification_item.dark.scss:1]: extra newline

File: addons/mail/static/src/core/common/primary_variables.scss [hunk 9]
```diff
@@ -31,5 +31,6 @@ $o-mail-Discuss-headerHeight: 48px !default;
 }
 
 .o-mail-systrayFullscreenDropdownMenu {
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/primary_variables.scss:33]: Not in diff but this looks miss-placed, it's not a primary variable.

File: addons/mail/static/src/core/web/messaging_menu.dark.scss [hunk 10]
```diff
@@ -0,0 +1,5 @@
+.o-mail-MessagingMenu-navbar button {
+    &.o-active, &:hover {
+        background-color: mix($o-gray-200, $o-gray-300);
```
Comment (seb-odoo) [inline addons/mail/static/src/core/web/messaging_menu.dark.scss:3]: Can't we settle on a color that we already have?

File: addons/mail/static/src/core/web/notification_item.dark.scss [hunk 11]
```diff
@@ -0,0 +1,9 @@
...
+.o-mail-NotificationItem {
```
Comment (seb-odoo) [inline addons/mail/static/src/core/web/notification_item.dark.scss:2]: comment not on the right line, but `/mail/static/src/img/smiley/mailfailure.jpg` doesn't work in dark theme because it has white background built-in

**Review by seb-odoo — APPROVED** [review]
Ok except comments, possibly missing tests to cover some non-style changes.

File: addons/mail/static/src/core/web/notification_item.dark.scss (hunk 11 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/web/notification_item.dark.scss:2]: Yes, this is a specific problem with the image using a white background instead of alpha like SMS failure does.
We should probably do a PR for this specific change

File: addons/mail/static/src/core/web/messaging_menu.dark.scss (hunk 10 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/web/messaging_menu.dark.scss:3]: Can't find one that has either similar color or semantics.

File: addons/mail/static/src/discuss/core/web/messaging_menu_patch.xml [hunk 12]
```diff
@@ -3,24 +3,20 @@
...
-                    </button>
+            <t t-if="ui.isSmall and (store.discuss.activeTab !== 'main' or !env.inDiscussApp)">
+                <div t-if="state.addingChat or state.addingChannel" class="w-100 p-1 border-bottom bg-light" t-ref="selector">
+                    <t t-if="state.addingChat" t-set="category" t-value="store.discuss.chats"/>
+                    <t t-if="state.addingChannel" t-set="category" t-value="store.discuss.channels"/>
+                    <ChannelSelector category="category" autofocus="true"/>
                 </div>
             </t>
+            <button t-if="!ui.isSmall or (ui.isSmall and store.discuss.activeTab !== 'channel' and !state.addingChat and !(env.inDiscussApp and store.discuss.activeTab === 'main'))" t-att-class="ui.isSmall ? 'w-100 p-2 btn btn-secondary border-bottom bg-light' : 'btn btn-link p-2'" t-on-click.stop="onClickNewMessage">
+                <t t-if="ui.isSmall">Start a conversation</t>
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/discuss/core/web/messaging_menu_patch.xml:14]: It looks off when there's no button on the livechat button, because the initial content is move up.

This button and the mailbox selection have the same height, ensuring the content part doesn't move up/down when changing tab.

One technique could be to keep same area as if the button is there but it's not (e.g. with `visibility: hidden`), but I think it's a better UX/UI to always have a button available.

Currently there's "New message"/"Start a conversation"/"New channel". The former 2 are the same except for label. It would be nice for follow-up work in master to have a single button for both "Start a conversation" and "New channel", as it feels best to have a single button that's always present that always do the same thing. There's already many changes of screens with the tabs and mailbox selection, the contextual buttons is too much and avoidable.

File: addons/mail/static/src/core/common/core.dark.scss [hunk 13]
```diff
@@ -1,3 +1,7 @@
 a.o_mail_redirect, a.o_channel_redirect {
     @include button-variant(rgba($primary, .2), rgba($primary, .2), lighten($link-color, 5%), rgba($primary, .3), rgba($primary, .3), lighten($link-color, 10%));
 }
+
+.o-mail-DiscussSystray-class.show {
+    background-color: $o-gray-100;
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/core/common/core.dark.scss:6]: Changed to `$white`. It's darker and better 🌚

File: addons/mail/static/src/core/common/core.scss [hunk 14]
```diff
@@ -51,3 +51,11 @@ a.o_mail_redirect, a.o_channel_redirect {
     @include button-size($btn-padding-y-sm, $btn-padding-x-sm, $btn-font-size-sm, $btn-border-radius-sm);
     padding: map-get($spacers, 1);
 }
+
+.o-mail-DiscussSystray {
+    --border-color: #{$o-gray-300} !important; // cancel custom border color of dropdown
+}
+
+.o-mail-DiscussSystray-class.show {
+    background-color: rgba(0, 0, 0, 0.075);
```
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/core/common/core.scss:60]: Indeed would be best as a systray concern. I don't feel confident to generalize the support in all systray items especially when nothing enforces them to use `Dropdown` toggler (it's at the preference of each systray item).

I have improved background effect to works with hover, and the background size takes the whole height (46px) of systray navbar.

File: addons/mail/static/src/core/common/primary_variables.scss (hunk 9 above)
Comment (alexkuhn) [reply addons/mail/static/src/core/common/primary_variables.scss:33]: Yeah conceptually it should be in `core.scss` (and all of its content, actually).

IIRC some variables need to be in `primary_variables` and not in `core.scss` (and vice-versa) due to some SCSS dependencies based on parsing of (S)CSS files at page load.

File: addons/mail/static/src/discuss/core/web/messaging_menu_patch.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/static/src/discuss/core/web/messaging_menu_patch.xml:13]: But starting a livechat conversation (from that menu in particular) makes no sense. And if it's creating a DM instead, it makes even less sense in the livechat menu.

**Review by seb-odoo — APPROVED** [review]
@robodoo r+

File: addons/mail/static/src/discuss/core/web/messaging_menu_patch.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (alexkuhn) [reply addons/mail/static/src/discuss/core/web/messaging_menu_patch.xml:13]: Probably the best in short-term is to remove the button in livechat tab.
This part of UI in mobile has ongoing tasks that will require more thought, e.g. having "Start a meeting" in the mobile screen. One idea was to put it in this header, but there's likely other and better options
